### PR TITLE
add Header-based canary routing via ingress annotations

### PR DESCRIPTION
**Pull request by witskeeper**
read  RouteHeader from ingress annotations, key: `yarp.ingress.kubernetes.io/route-headers`

### CODE REVIEW

**Comment by witskeeper** [comment]
#1254

File: src/Kubernetes.Controller/Converters/YarpIngressOptions.cs [hunk 1]
```diff
@@ -17,4 +17,24 @@ internal sealed class YarpIngressOptions
     public string CorsPolicy { get; set; }
     public HealthCheckConfig HealthCheck { get; set; }
     public Dictionary<string, string> RouteMetadata { get; set; }
+    public List<RouteHeader> RouteHeaders { get; set; }
 }
+
+internal sealed class RouteHeaderWapper
+{
+    public string Name { get; init; } = default!;
```
Comment (MihaZupan) [inline src/Kubernetes.Controller/Converters/YarpIngressOptions.cs:25]:
```suggestion
    public string Name { get; init; }
```

File: src/Kubernetes.Controller/Converters/YarpParser.cs [hunk 2]
```diff
@@ -202,7 +203,11 @@ private static YarpIngressOptions HandleAnnotations(YarpIngressContext context,
         {
             options.RouteMetadata = YamlDeserializer.Deserialize<Dictionary<string, string>>(routeMetadata);
         }
-
+        if (annotations.TryGetValue("yarp.ingress.kubernetes.io/route-headers", out var routeHeaders))
+        {
+            //YamlDeserializer does not support IReadOnlyList<string> in RouteHeader for now, so we use RouteHeaderWapper to solve this problem.
```
Comment (MihaZupan) [inline src/Kubernetes.Controller/Converters/YarpParser.cs:208]:
```suggestion
            // YamlDeserializer does not support IReadOnlyList<string> in RouteHeader for now, so we use RouteHeaderWapper to solve this problem.
```

File: test/Kubernetes.Tests/testassets/route-headers/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+metadata:
+  name: minimal-ingress
+  namespace: default
+  annotations:
+    yarp.ingress.kubernetes.io/route-metadata: |
+      foo: bar
+      another-key: another-value
+    yarp.ingress.kubernetes.io/route-headers: |
+      foo: bar
+      another-key: another-value
```
Comment (MihaZupan) [inline test/Kubernetes.Tests/testassets/route-headers/ingress.yaml:12]: Can you please add `route-headers` to the list of tests in https://github.com/microsoft/reverse-proxy/blob/3532c2b0ce9d4f95c42c86f3d658a9bb47915ca5/test/Kubernetes.Tests/IngressConversionTests.cs#L54

I would expect this to fail at the moment as this doesn't look like a valid format for the annotation, and doesn't match `routes.json`.

**Review by MihaZupan — COMMENTED** [review]
Thanks for working on this @witskeeper.

Can you please also add `route-headers` to the list of documented annotations here: https://github.com/microsoft/reverse-proxy/tree/main/samples/KubernetesIngress.Sample#annotations?

File: test/Kubernetes.Tests/testassets/route-headers/ingress.yaml (hunk 3 above)
Comment (witskeeper) [reply test/Kubernetes.Tests/testassets/route-headers/ingress.yaml:12]: OK，i will do it later

Comment (witskeeper) [reply test/Kubernetes.Tests/testassets/route-headers/ingress.yaml]: Done

**Review by MihaZupan — APPROVED** [review]

File: samples/KubernetesIngress.Sample/README.md [hunk 4]
```diff
@@ -70,6 +81,7 @@ The table below lists the available annotations.
 |yarp.ingress.kubernetes.io/route-metadata|Dictionary<string, string>|
 |yarp.ingress.kubernetes.io/session-affinity|[SessionAffinityConfig](https://microsoft.github.io/reverse-proxy/api/Yarp.ReverseProxy.Configuration.SessionAffinityConfig.html)|
 |yarp.ingress.kubernetes.io/transforms|List<Dictionary<string, string>>|
+|yarp.ingress.kubernetes.io/route-headers|List<[RouteHeaderWapper](/src/Kubernetes.Controller/Converters/YarpIngressOptions.cs#L23)>|
```
[the remark this answers is not in the document]
Comment (witskeeper) [reply samples/KubernetesIngress.Sample/README.md]: Done

**Comment by MihaZupan** [comment]
@JohannesLampel any thoughts on why the `license/cla` and `microsoft-cla` bots are in disagreement here?
Should we be ignoring the former?

**Comment by JohannesLampel** [comment]
@MihaZupan we've migrated to a new CLA app in late September. In that transition the CLA signing of @witskeeper was missed. I've updated everything in our database and the pr should be good to merge now.

**Comment by witskeeper** [comment]
Can this MR be merged？

**Comment by MihaZupan** [comment]
Thanks witskeeper

**Pull request merged**